src/mono/mono/metadata/class-init.c

```diff
@@ -1832,6 +1832,27 @@ type_has_references (MonoClass *klass, MonoType *ftype)
 	return FALSE;
 }
 
+static gboolean
+class_has_ref_fields (MonoClass *klass)
+{
+	/*
+	 * has_ref_fields is not set if this is called recursively, but this is not a problem since this is only used
+	 * during field layout, and instance fields are initialized before static fields, and instance fields can't
+	 * embed themselves.
+	 */
+	return klass->has_ref_fields;
+}
+
+static gboolean
+type_has_ref_fields (MonoType *ftype)
+{
+	if (m_type_is_byref (ftype) || class_has_ref_fields (mono_class_from_mono_type_internal (ftype)))
```

Comment:
```suggestion
	if (m_type_is_byref (ftype) || (mono_type_is_struct (ftype) && class_has_ref_fields (mono_class_from_mono_type_internal (ftype))))
```

@AaronRobinsonMSFT I think we want to guard the call to `mono_class_from_mono_type_internal` for reference types - it could trigger premature assembly loading, and also the field's class might not be fully initialized yet at this point, I think.

Also if we ever allow heap allocated objects to contain ref fields, the code seems wrong as-is: a struct S1 containing an object C that contains a struct S2 that has a ref field doesn't itself (S1) have ref fields, I think.

Reply:
I want an emoji that shows immense crying. I just spent the last 2.5 hours debugging mono to determine that.. Thanks. I confirmed the same.